File: backend/src/game-service-control.js
import { execFile } from "node:child_process";

export const GAME_CONTROL_HELPER_PATH =
  "/usr/local/sbin/novel-game-service-control";

const actionSet = new Set(["status", "start", "stop", "restart"]);
const statusSet = new Set([
  "running",
  "stopped",
  "failed",
  "activating",
  "deactivating",
  "unknown",
]);
const helperErrorSet = new Set([
  "game_control_action_failed",
  "game_control_action_invalid",
  "game_control_busy",
  "game_control_game_invalid",
  "game_control_not_root",
  "game_control_service_unavailable",
  "game_control_state_mismatch",
  "game_control_timeout",
]);
const unitGroups = Object.freeze({
  bailian: Object.freeze([
    "bailian-game.service",
    "bailian-backstage.service",
  ]),
  modao: Object.freeze([
    "modao-redis.service",
    "modao-static.service",
    "modao-account.service",
    "modao-game.service",
  ]),
});

export function unmanagedGameServiceStatus() {
  return {
    unit: "",
    status: "unmanaged",
    active: true,
    enabled: true,
    checkedAt: new Date().toISOString(),
    units: [],
  };
}

export function unavailableGameServiceStatus(gameId) {
  return {
    unit: String(gameId || ""),
    status: "unknown",
    active: false,
    enabled: false,
    checkedAt: new Date().toISOString(),
    units: [],
    error: "game_control_unavailable",
  };
}

export function runGameServiceControl(
  gameId,
  action,
  {
    execFileImpl = execFile,
    timeoutMs = 120_000,
  } = {},
) {
  const safeGameId = normalizeGameId(gameId);
  const safeAction = normalizeAction(action);
  return new Promise((resolve, reject) => {
    execFileImpl(
      "sudo",
      ["-n", GAME_CONTROL_HELPER_PATH, safeGameId, safeAction],
      {
        encoding: "utf8",
        timeout: Math.max(1_000, Math.min(180_000, Number(timeoutMs) || 120_000)),
        maxBuffer: 64 * 1024,
        windowsHide: true,
      },
      (error, stdout) => {
        const decoded = decodeHelperOutput(stdout);
        if (error || decoded?.ok !== true) {
          const code = helperErrorSet.has(decoded?.error)
            ? decoded.error
            : error?.killed
              ? "game_control_timeout"
              : "game_control_unavailable";
          reject(serviceError(code, code === "game_control_busy" ? 409 : 503));
          return;
        }
        try {
          resolve(sanitizeServiceStatus(safeGameId, decoded.data));
        } catch {
          reject(serviceError("game_control_unavailable", 503));
        }
      },
    );
  });
}

export function normalizeGameControlAction(value) {
  return normalizeAction(value);
}

function normalizeGameId(value) {
  const gameId = String(value || "").trim();
  if (!Object.hasOwn(unitGroups, gameId)) {
    throw serviceError("game_not_controllable", 400);
  }
  return gameId;
}

function normalizeAction(value) {
  const action = String(value || "").trim().toLowerCase();
  if (!actionSet.has(action)) {
    throw serviceError("game_action_invalid", 400);
  }
  return action;
}

function decodeHelperOutput(stdout) {
  try {
    const decoded = JSON.parse(String(stdout || "").trim());
    return decoded && typeof decoded === "object" ? decoded : null;
  } catch {
    return null;
  }
}

function sanitizeServiceStatus(gameId, raw) {
  if (!raw || typeof raw !== "object" || raw.gameId !== gameId) {
    throw new Error("invalid_helper_response");
  }
  const expectedUnits = unitGroups[gameId];
  if (!Array.isArray(raw.units) || raw.units.length !== expectedUnits.length) {
    throw new Error("invalid_helper_response");
  }
  const rawByUnit = new Map();
  for (const item of raw.units) {
    if (
      !item ||
      typeof item !== "object" ||
      typeof item.unit !== "string" ||
      rawByUnit.has(item.unit) ||
      !expectedUnits.includes(item.unit)
    ) {
      throw new Error("invalid_helper_response");
    }
    const status = String(item.status || "");
    if (!statusSet.has(status)) throw new Error("invalid_helper_response");
    rawByUnit.set(item.unit, {
      unit: item.unit,
      status,
      active: item.active === true,
      enabled: item.enabled === true,
      unitFileState: safeUnitFileState(item.unitFileState),
    });
  }
  const units = expectedUnits.map((unit) => rawByUnit.get(unit));
  if (units.some((unit) => !unit)) throw new Error("invalid_helper_response");
  return {
    unit: gameId,
    status: aggregateStatus(units),
    active: units.every((unit) => unit.active),
    enabled: units.every((unit) => unit.enabled),
    checkedAt: new Date().toISOString(),
    units,
  };
}

function safeUnitFileState(value) {
  const state = String(value || "").trim().toLowerCase();
  return /^[a-z][a-z0-9_-]{0,31}$/.test(state) ? state : "unknown";
}

function aggregateStatus(units) {
  if (units.every((unit) => unit.status === "running")) return "running";
  if (units.every((unit) => unit.status === "stopped")) return "stopped";
  if (units.some((unit) => unit.status === "failed")) return "failed";
  if (
    units.every((unit) =>
      ["running", "activating"].includes(unit.status),
    ) &&
    units.some((unit) => unit.status === "activating")
  ) {
    return "activating";
  }
  if (
    units.every((unit) =>
      ["stopped", "deactivating"].includes(unit.status),
    ) &&
    units.some((unit) => unit.status === "deactivating")
  ) {
    return "deactivating";
  }
  if (units.every((unit) => unit.status === "unknown")) return "unknown";
  return "partial";
}

function serviceError(code, statusCode) {
  const error = new Error(code);
  error.code = code;
  error.publicCode = code;
  error.statusCode = statusCode;
  return error;
}

File: backend/scripts/game-service-control.py
#!/usr/bin/env python3
import datetime
import json
import os
import subprocess
import sys

try:
    import fcntl
except ImportError:
    fcntl = None

SYSTEMCTL = "/usr/bin/systemctl"
LOCK_PATH = "/run/lock/novel-game-service-control.lock"
ALLOWED_ACTIONS = {"status", "start", "stop", "restart"}
SERVICE_GROUPS = {
    "bailian": (
        "bailian-game.service",
        "bailian-backstage.service",
    ),
    "modao": (
        "modao-redis.service",
        "modao-static.service",
        "modao-account.service",
        "modao-game.service",
    ),
    "kdjx": (
        "kdjx-mongodb.service",
        "kdjx-nsqlookupd.service",
        "kdjx-nsqd.service",
        "kdjx-host@accountdb.service",
        "kdjx-host@giftdb.service",
        "kdjx-host@storage1.service",
        "kdjx-host@storage2.service",
        "kdjx-host@pvp1.service",
        "kdjx-host@pvp2.service",
        "kdjx-host@crossdb.service",
        "kdjx-host@cross.service",
        "kdjx-anti-cheat.service",
        "kdjx-online-fight-forward.service",
        "kdjx-game@1.service",
        "kdjx-login.service",
        "kdjx-runtime.target",
    ),
}


def fail(code, exit_code=1):
    print(json.dumps({"ok": False, "error": code}, separators=(",", ":")))
    raise SystemExit(exit_code)


def systemctl(args, timeout=90):
    try:
        return subprocess.run(
            [SYSTEMCTL, *args],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=timeout,
            check=False,
            env={
                "PATH": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "LANG": "C",
                "LC_ALL": "C",
            },
        )
    except subprocess.TimeoutExpired:
        fail("game_control_timeout")
    except OSError:
        fail("game_control_service_unavailable")


def unit_status(unit):
    result = systemctl(
        [
            "show",
            unit,
            "--no-pager",
            "--property=LoadState",
            "--property=ActiveState",
            "--property=SubState",
            "--property=UnitFileState",
        ],
        timeout=15,
    )
    values = {}
    if result.returncode == 0:
        for line in result.stdout.splitlines():
            key, separator, value = line.partition("=")
            if separator:
                values[key] = value

    active_state = values.get("ActiveState", "unknown")
    load_state = values.get("LoadState", "unknown")
    unit_file_state = values.get("UnitFileState", "unknown").lower()
    if load_state != "loaded":
        status = "unknown"
    elif active_state == "active":
        status = "running"
    elif active_state == "inactive":
        status = "stopped"
    elif active_state == "failed":
        status = "failed"
    elif active_state == "activating":
        status = "activating"
    elif active_state == "deactivating":
        status = "deactivating"
    else:
        status = "unknown"

    return {
        "unit": unit,
        "status": status,
        "active": active_state == "active",
        "enabled": unit_file_state in {"enabled", "enabled-runtime"},
        "unitFileState": unit_file_state
        if unit_file_state.replace("-", "").replace("_", "").isalnum()
        else "unknown",
    }


def group_status(game_id, units):
    return {
        "gameId": game_id,
        "checkedAt": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        "units": [unit_status(unit) for unit in units],
    }


def acquire_action_lock():
    if fcntl is None:
        fail("game_control_service_unavailable")
    try:
        handle = open(LOCK_PATH, "a+", encoding="utf-8")
        os.chmod(LOCK_PATH, 0o600)
        fcntl.flock(handle.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
        return handle
    except BlockingIOError:
        fail("game_control_busy")
    except OSError:
        fail("game_control_service_unavailable")


def run_action(game_id, action, units):
    if action != "status":
        if action == "start":
            command = ["enable", "--now", *units]
        elif action == "stop":
            command = ["disable", "--now", *tuple(reversed(units))]
        else:
            command = ["restart", *units]
        result = systemctl(command)
        if result.returncode != 0:
            fail("game_control_action_failed")

    data = group_status(game_id, units)
    if action in {"start", "restart"} and not all(
        unit["active"] for unit in data["units"]
    ):
        fail("game_control_state_mismatch")
    if action == "start" and not all(
        unit["enabled"] for unit in data["units"]
    ):
        fail("game_control_state_mismatch")
    if action == "stop" and not all(
        unit["status"] == "stopped" and not unit["enabled"]
        for unit in data["units"]
    ):
        fail("game_control_state_mismatch")
    print(json.dumps({"ok": True, "data": data}, separators=(",", ":")))


def main():
    if os.geteuid() != 0:
        fail("game_control_not_root")
    if len(sys.argv) != 3:
        fail("game_control_action_invalid")
    game_id = sys.argv[1]
    action = sys.argv[2]
    units = SERVICE_GROUPS.get(game_id)
    if units is None:
        fail("game_control_game_invalid")
    if action not in ALLOWED_ACTIONS:
        fail("game_control_action_invalid")
    if not os.path.isfile(SYSTEMCTL) or not os.access(SYSTEMCTL, os.X_OK):
        fail("game_control_service_unavailable")
    lock_handle = acquire_action_lock() if action != "status" else None
    try:
        run_action(game_id, action, units)
    finally:
        if lock_handle is not None:
            lock_handle.close()


if __name__ == "__main__":
    main()
